File: src/schemas/knowledge_base.py
"""
Pydantic schemas for the Knowledge Base module.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, UUID4, field_validator, HttpUrl


class KnowledgeBaseCreate(BaseModel):
    name: str = Field(..., max_length=120)
    description: Optional[str] = Field(default=None, max_length=2000)
    language: Optional[str] = Field(default=None, max_length=16)
    embedding_model: Optional[str] = Field(default=None, max_length=255)
    chunk_size: int = Field(default=512, ge=64, le=4096)
    chunk_overlap: int = Field(default=128, ge=0, le=2048)
    config: Optional[Dict[str, Any]] = None


class KnowledgeBaseUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=120)
    description: Optional[str] = Field(default=None, max_length=2000)
    language: Optional[str] = Field(default=None, max_length=16)
    embedding_model: Optional[str] = Field(default=None, max_length=255)
    chunk_size: Optional[int] = Field(default=None, ge=64, le=4096)
    chunk_overlap: Optional[int] = Field(default=None, ge=0, le=2048)
    config: Optional[Dict[str, Any]] = None


class KnowledgeBaseResponse(BaseModel):
    id: UUID4
    client_id: UUID4
    name: str
    description: Optional[str]
    language: Optional[str]
    embedding_model: Optional[str]
    chunk_size: int
    chunk_overlap: int
    is_active: bool
    config: Dict[str, Any]
    created_by: Optional[UUID4]
    updated_by: Optional[UUID4]
    created_at: datetime
    updated_at: Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class KnowledgeDocumentCreate(BaseModel):
    source_type: str = Field(..., description="upload, url or text")
    original_filename: Optional[str] = Field(default=None, max_length=255)
    source_url: Optional[str] = Field(default=None, max_length=2048)
    mime_type: Optional[str] = Field(default=None, max_length=255)
    checksum: Optional[str] = Field(default=None, max_length=255)
    storage_path: Optional[str] = Field(default=None, max_length=1024)
    content_preview: Optional[str] = Field(default=None, max_length=4000)
    metadata: Optional[Dict[str, Any]] = None

    @field_validator("source_type")
    @classmethod
    def validate_source_type(cls, value: str) -> str:
        allowed = {"upload", "url", "text"}
        if value not in allowed:
            raise ValueError(f"source_type must be one of {', '.join(sorted(allowed))}")
        return value


class KnowledgeDocumentResponse(BaseModel):
    id: UUID4
    knowledge_base_id: UUID4
    client_id: UUID4
    source_type: str
    original_filename: Optional[str]
    source_url: Optional[str]
    mime_type: Optional[str]
    storage_path: Optional[str]
    checksum: Optional[str]
    content_preview: Optional[str]
    extra_metadata: Dict[str, Any]
    status: str
    error_message: Optional[str]
    created_by: Optional[UUID4]
    updated_by: Optional[UUID4]
    created_at: datetime
    updated_at: Optional[datetime]
    processing_started_at: Optional[datetime]
    processing_finished_at: Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class KnowledgeDocumentTextCreate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=255)
    content: str = Field(..., min_length=1, max_length=200000)


class KnowledgeDocumentUrlCreate(BaseModel):
    url: HttpUrl
    description: Optional[str] = Field(default=None, max_length=4000)

File: frontend/types/knowledgeBase.ts
/*
┌──────────────────────────────────────────────────────────────────────────────┐
│ @author: Davidson Gomes                                                      │
│ @file: /types/knowledgeBase.ts                                               │
│ Developed by: Davidson Gomes                                                 │
│ Creation date: May 13, 2025                                                  │
│ Contact: contato@evolution-api.com                                           │
├──────────────────────────────────────────────────────────────────────────────┤
│ @copyright © Evolution API 2025. All rights reserved.                        │
│ Licensed under the Apache License, Version 2.0                               │
│                                                                              │
│ You may not use this file except in compliance with the License.             │
│ You may obtain a copy of the License at                                      │
│                                                                              │
│    http://www.apache.org/licenses/LICENSE-2.0                                │
│                                                                              │
│ Unless required by applicable law or agreed to in writing, software          │
│ distributed under the License is distributed on an "AS IS" BASIS,            │
│ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.     │
│ See the License for the specific language governing permissions and          │
│ limitations under the License.                                               │
├──────────────────────────────────────────────────────────────────────────────┤
│ @important                                                                   │
│ For any future changes to the code in this file, it is recommended to        │
│ include, together with the modification, the information of the developer    │
│ who changed it and the date of modification.                                 │
└──────────────────────────────────────────────────────────────────────────────┘
*/

export type KnowledgeSourceType = "upload" | "url" | "text";

export interface KnowledgeBaseConfig {
  knowledge_base_ids?: string[];
  knowledge_base_context?: string;
  rag_top_k?: number;
  rag_score_threshold?: number;
  load_memory?: boolean;
  tools?: any[];
  mcp_servers?: any[];
  custom_mcp_servers?: any[];
  custom_tools?: any;
  sub_agents?: string[];
  agent_tools?: string[];
  max_iterations?: number;
  workflow?: any;
  tasks?: any[];
}

export interface KnowledgeBase {
  id: string;
  client_id: string;
  name: string;
  description?: string | null;
  language?: string | null;
  embedding_model?: string | null;
  chunk_size: number;
  chunk_overlap: number;
  is_active: boolean;
  config: Record<string, any>;
  created_by?: string | null;
  updated_by?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface KnowledgeBasePayload {
  name: string;
  description?: string;
  language?: string;
  embedding_model?: string;
  chunk_size?: number;
  chunk_overlap?: number;
  config?: Record<string, any>;
}

export interface KnowledgeDocument {
  id: string;
  knowledge_base_id: string;
  client_id: string;
  source_type: KnowledgeSourceType;
  original_filename?: string | null;
  source_url?: string | null;
  mime_type?: string | null;
  storage_path?: string | null;
  checksum?: string | null;
  content_preview?: string | null;
  extra_metadata: Record<string, any>;
  status: "pending" | "processing" | "ready" | "error";
  error_message?: string | null;
  created_by?: string | null;
  updated_by?: string | null;
  created_at: string;
  updated_at?: string | null;
  processing_started_at?: string | null;
  processing_finished_at?: string | null;
}

export interface KnowledgeDocumentTextPayload {
  title?: string;
  content: string;
}

export interface KnowledgeDocumentUrlPayload {
  url: string;
  description?: string;
}

export interface KnowledgeReference {
  document_id?: string;
  knowledge_base_id?: string;
  source?: string;
  chunk_index?: number;
  score?: number;
  metadata?: Record<string, any>;
}
